raphtory/src/python/graph/node_state/node_state.rs

```diff
@@ -504,64 +512,180 @@ impl_node_state_ord!(NodeStateGID<GID>, "NodeStateGID", "GID");
+        LazyNodeState::new(op, self.inner.nodes())
+    }
+
+    #[getter]
+    fn dt(
+        &self,
+    ) -> LazyNodeState<
+        'static,
+        ops::Map<ops::EarliestTime<DynamicGraph>, Result<Option<DateTime<Utc>>, TimeError>>,
+        DynamicGraph,
+        DynamicGraph,
+    > {
+        let op = self
+            .inner
+            .op
+            .clone()
+            .map(|t_opt| t_opt.map(|t| t.dt()).transpose());
+        LazyNodeState::new(op, self.inner.nodes())
+    }
+
+    #[getter]
+    fn secondary_index(
+        &self,
+    ) -> LazyNodeState<'static, EarliestSecondaryIndex<DynamicGraph>, DynamicGraph, DynamicGraph>
+    {
+        let op = self.inner.op.clone().map(|t_opt| t_opt.map(|t| t.i()));
+        LazyNodeState::new(op, self.inner.nodes())
+    }
+}
+type EarliestTimestamp<G> = ops::Map<ops::EarliestTime<G>, Option<i64>>;
+impl_lazy_node_state_ord!(
+    EarliestTimestampView<EarliestTimestamp<DynamicGraph>>,
     "NodeStateOptionI64",
     "Optional[int]"
 );
-impl_one_hop!(EarliestTimeView<ops::EarliestTime>, "EarliestTimeView");
-impl_node_state_group_by_ops!(EarliestTimeView, Option<i64>);
+impl_one_hop!(
+    EarliestTimestampView<EarliestTimestamp>,
+    "EarliestTimestampView"
+);
+impl_node_state_group_by_ops!(EarliestTimestampView, Option<i64>);
+
+type EarliestSecondaryIndex<G> = ops::Map<ops::EarliestTime<G>, Option<usize>>;
+impl_lazy_node_state_ord!(
+    EarliestSecondaryIndexView<EarliestSecondaryIndex<DynamicGraph>>,
+    "NodeStateOptionUsize",
+    "Optional[int]"
+); // usize gets converted to int in python
+impl_one_hop!(
+    EarliestSecondaryIndexView<EarliestSecondaryIndex>,
+    "EarliestSecondaryIndexView"
+);
+impl_node_state_group_by_ops!(EarliestSecondaryIndexView, Option<usize>);
+impl_node_state_ord!(
+    NodeStateOptionTimeIndexEntry<Option<TimeIndexEntry>>,
+    "NodeStateOptionTimeIndexEntry",
+    "Optional[TimeIndexEntry]"
+);
+impl_node_state_group_by_ops!(NodeStateOptionTimeIndexEntry, Option<TimeIndexEntry>);
+
 impl_lazy_node_state_ord!(
     LatestTimeView<ops::LatestTime<DynamicGraph>>,
     "NodeStateOptionI64",
     "Optional[int]"
 );
 impl_one_hop!(LatestTimeView<ops::LatestTime>, "LatestTimeView");
-impl_node_state_group_by_ops!(LatestTimeView, Option<i64>);
+impl_node_state_group_by_ops!(LatestTimeView, Option<TimeIndexEntry>);
+// implement custom functions for LazyNodeState<LatestTime>
+#[pymethods]
+impl LatestTimeView {
+    #[getter]
+    fn t(
+        &self,
+    ) -> LazyNodeState<'static, LatestTimestamp<DynamicGraph>, DynamicGraph, DynamicGraph> {
+        let op = self.inner.op.clone().map(|t_opt| t_opt.map(|t| t.t()));
+        LazyNodeState::new(op, self.inner.nodes())
+    }
+
+    #[getter]
+    fn dt(
+        &self,
+    ) -> LazyNodeState<
+        'static,
+        ops::Map<ops::LatestTime<DynamicGraph>, Result<Option<DateTime<Utc>>, TimeError>>,
+        DynamicGraph,
+        DynamicGraph,
+    > {
+        let op = self
```

Comment:
this should live in the rust layer, i.e., this should be able to just call self.inner.dt() here